pennylane/capture/flatfn.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Defines a utility for capturing higher order primitives that return pytrees.
+"""
+from functools import update_wrapper
+
+has_jax = True
+try:
+    import jax
+except ImportError:
+    has_jax = False
+
+
+# pylint: disable=too-few-public-methods
+class FlatFn:
+    """Wrap a function so that is accepts flat arguments and returns flat outputs. Caches
+    the pytree shape of the output into the ``out_tree`` property, so that the results can
+    be repacked later.
+
+    >>> def f(x):
+    ...     return {"y": 2+x["x"]}
+    >>> args = ({"x": 0},)
+    >>> flat_f = FlatFn(f)
+    >>> res = flat_f(*args)
+    >>> res
+    [2]
+    >>> jax.tree_util.tree_unflatten(flat_f.out_tree, res)
+    {'y': 2.5}
+
+    """
+
+    def __init__(self, f):
+        self.f = f
+        self.out_tree = None
+        update_wrapper(self, f)
+
+    def __call__(self, *args):
+        out = self.f(*args)
+        out_flat, out_tree = jax.tree_util.tree_flatten(out)
+        self.out_tree = out_tree
+        return out_flat
```

Comment:
So now the flattened function accepts unflattened inputs and gives flattened outputs?

Reply:
Could you update the docstring of `FlatFn` with info about the shape of the inputs and outputs. It could also be helpful to add some dev-specific context for why the function accepts unflat args, and also to note that evaluating the jaxpr will require flat inputs.